src/main/scala/mesosphere/marathon/core/storage/repository/impl/DeploymentRepositoryImpl.scala

```diff
@@ -97,4 +120,169 @@ class DeploymentRepositoryImpl[K, C, S](
+private class GcActor[K, C, S](
+  deploymentRepository: DeploymentRepositoryImpl[K, C, S],
+  groupRepository: StoredGroupRepositoryImpl[K, C, S],
+  appRepository: AppRepositoryImpl[K, C, S],
+  maxVersions: Int)(implicit mat: Materializer, ctx: ExecutionContext)
+    extends Actor with StrictLogging {
+  override val receive: Receive = idle
+
+  def idle: Receive = {
+    case DeletedPlan(plan) =>
+      delete()
+      context.become(deleting(Nil))
+  }
+
+  def deleting(pending: List[DeploymentPlan]): Receive = {
+    case DeletedPlan(plan) =>
+      context.become(deleting(plan :: pending))
+    case DeleteDone =>
+      if (pending.nonEmpty) {
+        delete()
+        context.become(deleting(Nil))
+      } else {
+        context.become(idle)
+      }
+  }
+
+  private def delete(): Unit = {
+    val future = async {
+      val rootVersions = await(groupRepository.rootVersions().runWith(Sink.sortedSet))
+      if (rootVersions.size <= maxVersions) {
+        Done
+      } else {
+        val currentRootFuture = groupRepository.root()
+        val storedPlansFuture = deploymentRepository.lazyAll().runWith(Sink.list)
+        val currentRoot = await(currentRootFuture)
+        val storedPlans = await(storedPlansFuture)
+
+        val inUseRootVersions: SortedSet[OffsetDateTime] = storedPlans.flatMap { plan =>
+          Seq(plan.originalVersion, plan.targetVersion)
+        }(collection.breakOut)
+
+        val deletionCandidates = rootVersions.diff(inUseRootVersions + currentRoot.version.toOffsetDateTime)
+
+        if (deletionCandidates.isEmpty) {
+          Done
+        } else {
+          val rootsToDelete = deletionCandidates.take(rootVersions.size - maxVersions)
+          if (rootsToDelete.isEmpty) {
+            Done
+          } else {
+            logger.info(s"Deleting Root Versions ${rootsToDelete.mkString(", ")} as there are currently " +
+              s"${rootVersions.size} which exceeds the maximum size $maxVersions and nothing refers to them anymore.")
+            // intentionally delete the roots first so apps can't be referred to.
+            await(Future.sequence(rootsToDelete.map(groupRepository.deleteRootVersion)))
+            await(deleteUnusedApps(storedPlans, currentRoot))
+          }
+        }
+      }
+    }
+    future.onFailure {
+      case NonFatal(e) => logger.error("Failed to delete extra root versions", e)
+    }
+    future.onComplete(_ => self ! DeleteDone)
+  }
+
+  private def deleteUnusedApps(
+    storedPlans: Seq[StoredPlan],
+    currentRoot: Group): Future[Done] = {
+
+    def appsInUse(roots: Seq[StoredGroup]): Map[PathId, SortedSet[OffsetDateTime]] = {
+      val appVersionsInUse = new mutable.HashMap[PathId, mutable.Set[OffsetDateTime]] with mutable.MultiMap[PathId, OffsetDateTime] // scalastyle:off
+      currentRoot.transitiveAppsById.foreach {
+        case (id, app) =>
+          appVersionsInUse.addBinding(id, app.version.toOffsetDateTime)
+      }
+      roots.foreach { root =>
```

Comment:
That one is actually pretty funny - currentRoot is a `Group` while roots is a collection of `StoredGroup`. While very similar, they aren't the same and I was trying to avoid creating an extra `StoredGroup`.